roles/scripts/files/nightking.bash

```diff
@@ -27,58 +37,73 @@ do
     fi
-    stemplate /etc/experiments/"${XP}"/seed/ -o /home/tendermint/.tendermint/ -f /etc/experiments/"${XP}"/config.toml --all
+    stemplate "${XP_DIR}"/seed/ -o /home/tendermint/.tendermint/ -f "${XP_DIR}"/config.toml --all
     chown -R tendermint.tendermint /home/tendermint/.tendermint
-    sudo -u tendermint tendermint show_node_id > /var/log/nightking/seed_node_id
-    export NIGHTKING_SEED_NODE_ID="$(cat /var/log/nightking/seed_node_id)"
+    sudo -u tendermint tendermint show_node_id > "${CACHE_DIR}/nightking-seed-node-id"
+    export NIGHTKING_SEED_NODE_ID="$(get nightking-seed-node-id)"
     systemctl start tendermint
     log seed 0
     trap '' ERR
 
+    ##
     # Set up terraform
-    trap 'log terraform_build 12' ERR
-        export SSH_KEY="$(tail -1 /home/ec2-user/.ssh/authorized_keys)"
-    if [ -f /etc/experiments/"${XP}"/config.toml ]; then
-      stemplate /usr/share/terraform-templates --env -f /etc/experiments/"${XP}"/config.toml -o /root/terraform-"${XP}" --all
-      if [ -d /etc/experiments/"${XP}"/terraform ]; then
-        stemplate /etc/experiments/"${XP}"/terraform --env -f /etc/experiments/"${XP}"/config.toml -o /root/terraform-"${XP}" --all
+    ##
+    PREVIOUS_INFRA="$(peek-cache infra || echo '')"
+    # Build a new infra or ...
+    if [ -z "${PREVIOUS_INFRA}" ]; then
+      trap 'log terraform_build 12' ERR
+      export SSH_KEY="$(tail -1 /home/ec2-user/.ssh/authorized_keys)"
+      if [ -f "${XP_DIR}"/config.toml ]; then
+        stemplate /usr/share/terraform-templates --env -f "${XP_DIR}"/config.toml -o /root/terraform-"${XP}" --all
+        if [ -d "${XP_DIR}"/terraform ]; then
+          stemplate "${XP_DIR}"/terraform --env -f "${XP_DIR}"/config.toml -o /root/terraform-"${XP}" --all
+        fi
+      else
+        if [ -d "${XP_DIR}"/terraform ]; then
+          stemplate "${XP_DIR}"/terraform -o /root/terraform-"${XP}" --env -f "${XP_DIR}"/config.toml --all
+        fi
       fi
+      test -d /root/terraform-"${XP}" || echo "Experiment folder /root/terraform-${XP} not found."
+      test -d /root/terraform-"${XP}" || continue
+      cd /root/terraform-"${XP}"
+      trap 'log terraform_build 11' ERR
+      log terraform_build 2
+      terraform init
+      trap 'log terraform_build 10' ERR
+      log terraform_build 1
+      export TERRAFORM_RESULT=0
+      terraform apply --auto-approve -no-color || export TERRAFORM_RESULT=$?
+    # ... reinitialize a previous infra.
     else
-      if [ -d /etc/experiments/"${XP}"/terraform ]; then
-        stemplate /etc/experiments/"${XP}"/terraform -o /root/terraform-"${XP}" --env -f /etc/experiments/"${XP}"/config.toml --all
-      fi
+      IP_LIST="$(terraform show /root/terraform-${PREVIOUS_INFRA}/terraform.tfstate -no-color | grep "^  public_ip =" | sed "s/^  public_ip = //" | tr '\n' ' ')"
+      sudo -u ec2-user pssh -p 20 -t 10 -i -O "StrictHostKeyChecking no" -H "${IP_LIST}" sudo nohup /usr/local/sbin/run-role.bash "${XP}" &
```

Comment:
Nice! 🚀  So just to clarify, this checks whether the network's already been deployed by Terraform: if not, it deploys it (which automatically runs the `run-role.bash` script); if it has already been deployed, it uses `pssh` to execute `run-role.bash` on the target machines?

Reply:
Yes. This was recently clarified more in the script by pulling some of the code into functions. `run-bash` was renamed to `runxp` for easier typing.